### DISCUSSION TITLE
Networked control / API /CLI support

### DISCUSSION THREAD

**Discussion by waydabber** [MonitorControl/MonitorControl#682] [Q&A] [answered]
### Before opening the issue, have you...?

- [X] Searched for existing issues

### Is your feature request related to a problem? Please describe

It would be great to add a simple client/server mode to the app + define a simple api and add an option to add devices via network control using various networked protocols (this is related to #287 as well). This could enable control of multiple mac displays from a single screen (which might not be a very convincing use case) and more importantly allow for AirPlay connected macs (used as external displays on Monterey as Macs now can basically act as Sidecars to each other, like the old Target Display mode) to have brightness control (which is a more convincing use case - let's think about all those 4K/5K Intel iMacs which will be obsolete within a few years but could still be great external displays for an Apple Silicon mac with a wired or wireless AirPlay connection).

### Describe the solution you'd like

A simple option to enable listening on a port for control + an option to add a networked device on the 'Displays' tab (possibly renamed to 'Devices' in the future). Everything else should work the way it is working now, so the implementation would need any serious overhaul of the UI.

### Describe alternatives you've considered

Not doing this.

**Comment by JoniVR** [comment]
I like the idea but I wonder if it's not a bit out-of-scope? 
(Do few things and do them well vs do-it-all)

Would also possibly bring a few issues that would be purely related to ports being blocked, firewall settings, ...?

Still, interesting idea!

**Comment by waydabber** [comment]
Well, I just popped it, I don't think it will be implemented any time soon. I'll convert this to a discussion with an 'Idea' tag on it and see if it gains traction.

**Comment by Kane610** [comment]
One reason why a network API would be nice here is to be able to automate control of the displays from an automation system such as Home Assistant.

**Reply by waydabber** [reply]
Hmm. Didn't think of that but you are right.

**Reply by the0neyouseek** [reply]
I think a simple cli api would suffice for this use case

**Reply by Kane610** [reply]
CLI wouldn't work well with most automation systems since they typically don't run on a client machine. If automation systems are relevant to support a network API would be the way.

**Reply by JoniVR** [reply]
@the0neyouseek For stuff like home assistant probably not, because you usually run that on a NAS in docker. Though users who really want it can probably script it pretty easily with already available tools. A CLI would just simplify that scripting process a bit, but not as much as an API would.

It's a very valid use-case for a network api, question is just a matter of: 
"Are we going to support it if only a handful of people will ever use it?"

**Reply by Kane610** [reply]
Its also good from a marketing stand point if you want to be more visible to new users ;)

**Reply by the0neyouseek** [reply]
Doesn't home assistant allows you to send command over ssh ?

**Reply by Kane610** [reply]
There is not an integration for it, so it would depend on what the host system has installed. Which typically is some sort of containerised environment.

**Reply by fbirlik** [reply]
Also let's not forget about security implications of such a network endpoint.  For company laptops, IT people may not mind users installing a brightness control app, but they may be more reluctant if the same app has a potential remote access mechanism attached to it.

What about authentication and access control?  MonitorControl has 'accessibility' privileges, so a potential intruder may access more than just the brightness control.

Having said that, I believe a local cli and some additional guide or maybe even an additional helper application that would 'networkify' the cli would keep these problems at bay.

**Comment by faroit** [comment]
I just want to throw in that a very simple cli interface would be quite useful. For the internal display in tended to use https://github.com/nriley/brightness and connect that to Alfred or some other launcher.

**Reply by pellet** [reply]
Integrating this with the existing brightness CLI would be awesome.

**Reply by HenryGetz** [reply]
Came here to second this.... this would allow for super powerful workflows when combined with something like Better Touch Tool

**Comment by waydabber** [comment, marked as answer]
Hi all, please use https://github.com/waydabber/BetterDisplay + https://github.com/waydabber/betterdisplaycli. BetterDisplay provides both CLI, HTTP, URL Schema and notification dispatch integration. A similar feature probably won't be part of MonitorControl ever.

https://github.com/waydabber/BetterDisplay/wiki/Integration-features,-CLI
